Makefile

```diff
@@ -1,36 +1,44 @@
 # vim: tabstop=8 shiftwidth=8 noexpandtab:
 
-all: mimiker.elf tags cscope initrd.cpio
+all: cscope tags mimiker.elf initrd.cpio
 
 # Disable all built-in recipes
 .SUFFIXES:
 
 include Makefile.common
 $(info Using CC: $(CC))
 
+# Directories which contain kernel parts
+SYSSUBDIRS  = mips stdc sys tests
 # Directories which require calling make recursively
-SUBDIRS = mips stdc sys user tests
+SUBDIRS = $(SYSSUBDIRS) user
 
-# This rule ensures that all subdirectories are processed before any file
-# generated within them is used for linking the main kernel image.
-$(KRT): | $(SUBDIRS)
-	true # Disable default recipe from Makefile.common
+$(SUBDIRS):
+	$(MAKE) -C $@
+
+.PHONY: format tags cscope $(SUBDIRS) force
+
+# Make sure the global cache dir exists before building user programs
+user: | cache
+cache:
+	mkdir cache
 
-mimiker.elf: $(KRT)
+mimiker.elf: $(KRT) | $(SYSSUBDIRS)
 	@echo "[LD] Linking kernel image: $@"
 	$(CC) $(LDFLAGS) -Wl,-Map=$@.map $(LDLIBS) -o $@
 
 cscope:
 	cscope -b include/*.h ./*/*.[cS]
 
 tags:
-	find -iname '*.[ch]' -not -path "*/toolchain/*" | ctags --language-force=c -L-
-	find -iname '*.[ch]' -not -path "*/toolchain/*" | ctags --language-force=c -L- -e -f etags
-	find -iname '*.S' -not -path "*/toolchain/*" | ctags -a --language-force=asm -L-
-	find -iname '*.S' -not -path "*/toolchain/*" | ctags -a --language-force=asm -L- -e -f etags
+	@echo "Rebuilding tags..."
+	find -iname '*.[ch]' -not -path "*/toolchain/*" -and -not -path "*newlib*" | ctags --language-force=c -L-
```

Comment:
What about introducing additional step that generates `sources.lst` file and then reuses it to run `ctags`?